Remove the need for extra connections between agent and manager.

thallgren opened the pull request:
## Description

Replaces the forwarding of intercepted connections from agent to the
manager with a ConnTunnel so that no extra connections have to be made
to ports that aren't exposed in the manager's deployment manifest.

Also ensures that all outbound traffic from the client to the
traffic-manager is routed to intercepted agents so that those requests
originates from a pod in the currently intercepted namespace.

A welcome side-effect is that the agent no longer needs to export ports
externally in order to serve remote VolumeMounts since those connections
too originates in the agent.

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [x] My change is adequately tested.

Commit:
Remove listener.go. It's no longer used.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move context sessionInfo methods to managerutil package

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add manager.AgentTunnel. Rename ConnTunnel to ClientTunnel

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Remove the need for extra connections between agent and manager.

Replaces the forwarding of intercepted connections from agent to the
manager with a ConnTunnel so that no extra connections have to be made
to ports that aren't exposed in the manager's deployment manifest.

Also ensures that all outbound traffic from the client to the
traffic-manager is routed to intercepted agents so that those requests
originates from a pod in the currently intercepted namespace.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update CHANGELOG.md

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

LukeShu commented:
One day I will find out the reason why v4 ICMP and v6 ICMP have different numbers.

LukeShu commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
 # Changelog
 
+### 2.3.0 TBD
```
2.3.1

LukeShu commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
+### 2.3.0 TBD
+
+- Change: Outbound connections are now routed to intercepted agents.
+- Change: Inbound connection from intercepted agent now uses existing gRPC instead of new connection to manager port between 6000-7999
```
This sentence reads super weird to me.  Perhaps:

> - Change: Inbound connections from an intercepted agent are now tunneled to the manager over the existing gRPC connection, instead of establishing a new connection to the manager for each inbound connection.

LukeShu commented:
If I were the one doing this, I'd squash the last 3 commits together.

LukeShu commented on cmd/traffic/cmd/agent/forwarder.go:
```diff
@@ -91,34 +116,74 @@ func (f *Forwarder) Target() (string, int32) {
 	if err != nil {
-		dlog.Infof(ctx, "Error on resolve(%s:%d): %+v", targetHost, targetPort, err)
-		return
+		return fmt.Errorf("error on resolve(%s:%d): %+v", targetHost, targetPort, err)
```
Use `%w` when wrapping errors.

LukeShu commented on cmd/traffic/cmd/agent/forwarder.go:
```diff
@@ -131,8 +196,7 @@ func (f *Forwarder) ForwardConn(clientConn *net.TCPConn) {
 	if err != nil {
-		dlog.Infof(ctx, "Error on dial: %+v", err)
-		return
+		return fmt.Errorf("error on dial: %+v", err)
```
Likewise.

LukeShu commented on cmd/traffic/cmd/agent/forwarder.go:
```diff
@@ -157,8 +221,83 @@ func (f *Forwarder) ForwardConn(clientConn *net.TCPConn) {
 	for i := 0; i < 2; i++ {
 		select {
 		case <-ctx.Done():
-			return
 		case <-done:
 		}
 	}
```
I almost posted a comment about how the original code was wrong, but that was me mis-interpretting the code.

The original code would have (IMO) been more clearly written as
```go
	// Wait for both sides to close the connection
	for numClosed := 0; numClosed < 2; {
		select {
		case <-ctx.Done():
			return
		case <-done:
			numClosed++
		}
	}
```

That is:
 1. In the "happy" path, we don't want to terminate until `<-done` has been read twice, or
 2. In the "sad" path, we want to bail early if `ctx.Done()` gets closed.
 
With that in mind, I'm not sure how it's correct to remove the bail-early `return`.

But I also don't see how it's correct to orphan the goroutines as it did when bailing early.  It seems to me that the reason `done` is a channel at all (instead of a `sync.WaitGroup`) is in order to support context cancellation.

LukeShu commented:
I get the feeling that we're acquiring a lot of duplicate code around pumping forwarded connections (manager/systema, manager/agent, agent/app...). I might be adding "deduplicate all that" to my cooldown list.

LukeShu commented:
I guess part of that is that in my mind the connpool thing works that way, but really it doesn't, since it's handling multiplexing.  Which is a little silly, because the HTTP/2 stream that gRPC is using can already handle multiplexing.

LukeShu commented:
Minor concern/question about cancellation in the agent.  A couple of nitpicks.  Otherwise, looks good.

Commit:
Changes in response to review

- Change version and wording of changlog entry
- Use %w when wrapping errors
- Clearer version of waiting for two sides of connection to close

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

LukeShu reviewed (approved):
Hit it!

Commit:
Split Serve method into Listen and ServerListener.

The change enables access to the listener so that its port number can
be extracted before the actual accept loop begins.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix wrong directory of arrow in sendTCD()

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

LukeShu committed:
Merge branch 'release/v2' into thallgren/mesh-intergration
